### PR TITLE
Implement listen_close notification and tests

### PR DESCRIPTION
**Pull request by bomanaps**
## What was wrong?

Issue #https://github.com/libp2p/py-libp2p/issues/825

## How was it fixed?
This PR completes the core “listen_close” functionality and ensures tests now validate's it.

Summary of approach.
Extended the notification system with additional tests and functionality, covering both listen_close and closed_stream events

# What’s completed in this PR:
- [x] Implement listen_close in MyNotifee (#827)
- [x] Implement swarm notification for listener closure (notify_listen_close) (#828)
- [x] Add integration points for listener lifecycle (e.g. calling notify_listen_close) (#829)
- [x] Write tests for listen_close (both normal and edge cases) (#830)

### To-Do

- [ ] Clean up commit history
- [ ] Add or update documentation related to these changes
- [ ] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by bomanaps** [comment]
Hello @acul71 please can you review, added extra test and edge cases

File: libp2p/network/swarm.py [hunk 1]
```diff
@@ -330,8 +330,18 @@ async def close(self) -> None:
 
             # Close all listeners
             if hasattr(self, "listeners"):
-                for listener in self.listeners.values():
+                for maddr_str, listener in self.listeners.items():
                     await listener.close()
+                    # Notify about listener closure
+                    try:
+                        from multiaddr import Multiaddr
```
Comment (acul71) [inline libp2p/network/swarm.py]: Is there a reason for this import to be here (not on top)?

**Comment by acul71** [comment]
@bomanaps 
Hey! 👋 

So this PR implements the `listen_close` notification functionality and adds comprehensive tests for the notification system. You've implemented both `listen_close` and `closed_stream` notifications, plus added a bunch of new test cases. Overall, it's pretty solid!

## What I found 🔍

### Good stuff:
- **Implemented `notify_listen_close`** - replaced the `NotImplementedError` with actual implementation
- **Added listener closure integration** - now properly notifies when listeners are closed during swarm shutdown
- **Implemented `listen_close` in MyNotifee** - added proper parameter validation and event tracking
- **Added comprehensive tests** - 5 new test functions covering various scenarios
- **Error handling** - added proper exception handling for notification failures

### One small issue:
There's a local import that should be moved to the top of the file:

```python
# Line 336 in swarm.py - this is redundant:
try:
    from multiaddr import Multiaddr  # <- Remove this line
    multiaddr = Multiaddr(maddr_str)
    await self.notify_listen_close(multiaddr)

# Just use the top-level import that's already there:
try:
    multiaddr = Multiaddr(maddr_str)
    await self.notify_listen_close(multiaddr)
```

**Note**: This is the only real issue I found - everything else looks good!

## Test coverage thoughts 🤔

You added a lot of great tests! You've got:
- ✅ **5 new test functions** covering different scenarios
- ✅ Error handling for invalid parameters in `listen_close` and `closed_stream`
- ✅ Rapid stream operations testing
- ✅ Concurrent operations with Trio nursery
- ✅ Edge cases for the notification system

The test coverage is actually quite comprehensive for what you implemented. The only things that could be added in a future PR:
- ❌ Real `listen`/`listen_close` testing during actual listener lifecycle (not just cleanup)
- ❌ `PubsubNotifee` integration tests
- ❌ Performance testing with many notifees

## My take 💭

**Approved with minor changes**. The core implementation is good and the tests cover the essentials. The import thing is a quick fix, and while more test coverage would be nice, it's not blocking for merge.

Just fix that import and you're good to go! 🚀

**Note**: Would be great to see a follow-up PR with additional test coverage for the missing scenarios I mentioned above - especially the `PubsubNotifee` integration tests and failure scenarios. But that can be a separate improvement PR.

**Comment by bomanaps** [comment]
> @bomanaps Hey! 👋
> 
> So this PR implements the `listen_close` notification functionality and adds comprehensive tests for the notification system. You've implemented both `listen_close` and `closed_stream` notifications, plus added a bunch of new test cases. Overall, it's pretty solid!
> ## What I found 🔍
> ### Good stuff:
> 
>     * **Implemented `notify_listen_close`** - replaced the `NotImplementedError` with actual implementation
> 
>     * **Added listener closure integration** - now properly notifies when listeners are closed during swarm shutdown
> 
>     * **Implemented `listen_close` in MyNotifee** - added proper parameter validation and event tracking
> 
>     * **Added comprehensive tests** - 5 new test functions covering various scenarios
> 
>     * **Error handling** - added proper exception handling for notification failures
> 
> 
> ### One small issue:
> 
> There's a local import that should be moved to the top of the file:
> 
> ```python
> # Line 336 in swarm.py - this is redundant:
> try:
>     from multiaddr import Multiaddr  # <- Remove this line
>     multiaddr = Multiaddr(maddr_str)
>     await self.notify_listen_close(multiaddr)
> 
> # Just use the top-level import that's already there:
> try:
>     multiaddr = Multiaddr(maddr_str)
>     await self.notify_listen_close(multiaddr)
> ```
> 
> **Note**: This is the only real issue I found - everything else looks good!
> ## Test coverage thoughts 🤔
> 
> You added a lot of great tests! You've got:
> 
>     * ✅ **5 new test functions** covering different scenarios
> 
>     * ✅ Error handling for invalid parameters in `listen_close` and `closed_stream`
> 
>     * ✅ Rapid stream operations testing
> 
>     * ✅ Concurrent operations with Trio nursery
> 
>     * ✅ Edge cases for the notification system
> 
> 
> The test coverage is actually quite comprehensive for what you implemented. The only things that could be added in a future PR:
> 
>     * ❌ Real `listen`/`listen_close` testing during actual listener lifecycle (not just cleanup)
> 
>     * ❌ `PubsubNotifee` integration tests
> 
>     * ❌ Performance testing with many notifees
> 
> 
> ## My take 💭
> 
> **Approved with minor changes**. The core implementation is good and the tests cover the essentials. The import thing is a quick fix, and while more test coverage would be nice, it's not blocking for merge.
> 
> Just fix that import and you're good to go! 🚀
> 
> **Note**: Would be great to see a follow-up PR with additional test coverage for the missing scenarios I mentioned above - especially the `PubsubNotifee` integration tests and failure scenarios. But that can be a separate improvement PR.

I will open a fellow up PR to address this.

**Review by acul71 — APPROVED** [review]

**Comment by acul71** [comment]
@pacrob @sumanjeet0012 
This PR has been reviewed and ready to merged. 
This PR completes (more tests will be added in next PR) 
The notification system in py-libp2p is the **nervous system** of the peer-to-peer network, allowing different components to stay synchronized with network state changes without direct dependencies. It uses an event-driven architecture where components can register as "notifees" to receive notifications about network lifecycle events.
It helps Peer *Discovery and Management*, *Resource Management*, *Listener Lifecycle Management*, *Debugging and Monitoring*
More info here: https://github.com/libp2p/py-libp2p/discussions/724#discussioncomment-14129320

**Comment by seetadev** [comment]
@bomanaps , @acul71 : Great work on this PR! 🎉 The way you’ve described the notification system as the “nervous system” of py-libp2p is both accurate and elegant—it really highlights the critical role it plays in keeping the network state coherent while maintaining modularity.

I particularly appreciate how this PR ties together multiple aspects of the notification lifecycle:

- listen_close implementation in MyNotifee 

- swarm notification hooks for both closed_stream and listen_close 

- integration across stream and listener lifecycle 

and the thoughtful tests for both normal and edge cases 

This is a strong, end-to-end approach that doesn’t just deliver the functionality but also ensures its correctness and robustness. The test coverage is especially important here, given how subtle and interconnected lifecycle events can be.

Wish if we could add some more explanation on the discussion page related to the implementation of the PR for other new cohort members.

The PR description is super helpful for newcomers—it makes it clear why the notification system matters and how these changes fit into the bigger picture of peer discovery, resource management, and debugging.

Thanks again for pushing this forward—this is a meaningful step in strengthening py-libp2p’s event-driven architecture. 

Looking forward to the follow-up PRs that extend test coverage even further.
